### PR TITLE
Krill MUST NOT create only a single AFI ASPA #1063

### PR DESCRIPTION
**Pull request by timbru**

### CODE REVIEW

File: src/commons/api/aspa.rs [hunk 1]
```diff
@@ -100,70 +100,28 @@ impl AspaDefinition {
...
-                                if removed_limit == existing_limit {
-                                    // there is a match to remove
-                                    return true;
-                                }
-                                // else -> there is no match to remove
-                            }
-                        }
-                    }
-                }
+    /// Returns true if this is a single AFI AspaDefinition. Technically,
```
Comment (partim) [inline src/commons/api/aspa.rs:103]: Your doc string and the method name contradict each other – the doc string suggests this returns true if there is only one AFI, the method name suggests it returns true if it has both AFIs.

**Review by partim — CHANGES_REQUESTED** [review]

File: src/commons/api/aspa.rs (hunk 1 above)
Comment (ximon18) [inline src/commons/api/aspa.rs:103]: The fn name says "both" but the comment says "single" - this sounds contradictory.

File: src/commons/api/aspa.rs [hunk 2]
```diff
@@ -100,70 +100,28 @@ impl AspaDefinition {
...
+    /// it is allowed to omit one address family entirely, but this will
+    /// be interpreted as thought the user specified an AS0 provider for
```
Comment (ximon18) [inline src/commons/api/aspa.rs:105]: s/as thought/as though/

File: src/daemon/ca/certauth.rs [hunk 3]
```diff
@@ -1841,21 +1845,40 @@ impl CertAuth {
...
-    /// Verifies whether the update needs to be applied
+    /// Verifies whether the update is allowed and needs to be applied.
     ///
     /// The update does not need to be applied if there would be no change in
     /// the configured AspaDefinition. I.e. this gives us idempotence and e.g. allows
     /// an operator just issue a command to add a provider for a customer ASN, and
     /// if it was already authorised then no work is needed.
-    fn updated_needed(&self, customer: AspaCustomer, update: &AspaProvidersUpdate) -> KrillResult<bool> {
-        if update.is_empty() {
+    fn updated_allowed_and_needed(&self, customer: AspaCustomer, update: &AspaProvidersUpdate) -> KrillResult<bool> {
```
Comment (ximon18) [inline src/daemon/ca/certauth.rs:1854]: s/updated/update/ ?

File: src/daemon/ca/certauth.rs [hunk 4]
```diff
@@ -1841,21 +1845,40 @@ impl CertAuth {
         Ok(update_events)
     }
 
-    /// Verifies whether the update needs to be applied
+    /// Verifies whether the update is allowed and needs to be applied.
     ///
     /// The update does not need to be applied if there would be no change in
     /// the configured AspaDefinition. I.e. this gives us idempotence and e.g. allows
     /// an operator just issue a command to add a provider for a customer ASN, and
```
Comment (ximon18) [inline src/daemon/ca/certauth.rs:1852]: s/an operator just/an operator to just/

Comment (ximon18) [reply src/daemon/ca/certauth.rs:1852]: Ah, ignore.

File: src/commons/api/aspa.rs (hunk 1 above)
Comment (timbru) [reply src/commons/api/aspa.rs:103]: yeah, oops sorry.. refactored and then did not change this text

Comment (timbru) [reply src/commons/api/aspa.rs:103]: yeah, oops sorry.. refactored and then did not change this text

File: src/daemon/ca/certauth.rs (hunk 3 above)
Comment (timbru) [reply src/daemon/ca/certauth.rs:1854]: The function argument is the 'update', but the result is (potentially) something 'updated'. So, leaving this as-is.

File: src/commons/api/aspa.rs (hunk 1 above)
Comment (timbru) [reply src/commons/api/aspa.rs]: fixed

Comment (timbru) [reply src/commons/api/aspa.rs]: fixed

File: src/commons/api/aspa.rs (hunk 2 above)
Comment (timbru) [reply src/commons/api/aspa.rs]: fixed

**Comment by timbru** [comment]
> I can't speak to whether or not other parts of the code base should be changed too or if this makes sense from a business logic point of view.

Discussion is still ongoing, but this seems safest for now.

This does not change existing ASPAs as far as they have been defined, but prevents new instances or changes that would result in this.

**Review by partim — APPROVED** [review]

**Pull request merged**